src/extension/infer.rs

```diff
@@ -546,57 +546,50 @@ impl UnificationContext {
     pub fn results(&self) -> Result<ExtensionSolution, InferExtensionError> {
         // Check that all of the metavariables associated with nodes of the
         // graph are solved
+        let depended_upon = {
+            let mut h: HashMap<Meta, Vec<Meta>> = HashMap::new();
+            for (m, m2) in self.constraints.iter().flat_map(|(m, cs)| {
+                cs.iter().flat_map(|c| match c {
+                    Constraint::Plus(_, m2) => Some((*m, self.resolve(*m2))),
+                    _ => None,
+                })
+            }) {
+                h.entry(m2).or_default().push(m);
+            }
+            h
+        };
+        // Calculate everything dependent upon a variable.
+        // Note it would be better to find metas ALL of whose dependencies were (transitively)
+        // on variables, but this is more complex, and hard to define if there are cycles
+        // of PLUS constraints, so leaving that as a TODO until we've handled such cycles.
+        let mut depends_on_var = HashSet::new();
+        let mut queue = VecDeque::from_iter(self.variables.iter());
+        while let Some(m) = queue.pop_front() {
+            if depends_on_var.insert(m) {
+                if let Some(d) = depended_upon.get(m) {
+                    queue.extend(d.iter())
+                }
+            }
+        }
+
         let mut results: ExtensionSolution = HashMap::new();
         for (loc, meta) in self.extensions.iter() {
             if let Some(rs) = self.get_solution(meta) {
                 if loc.1 == Direction::Incoming {
                     results.insert(loc.0, rs.clone());
                 }
-            } else if self.live_var(meta).is_some() {
+            } else if !depends_on_var.contains(&self.resolve(*meta)) {
                 // If it depends on some other live meta, that's bad news.
                 return Err(InferExtensionError::Unsolved { location: *loc });
             }
-            // If it only depends on graph variables, then we don't have
+            // If it ("only"??) depends on graph variables, then we don't have
```

Comment:
This only is what bothers me; the above calculation includes things that depend on both graph variables and other things.

See branch `fix/infer_stack_overflow2` for a simpler check that things depend upon solved values are solved; but (step through the commits) this is basically pointless.

See branch `fix/infer_stack_overflow3` for a more-complex check. Upon reflection I think maybe that is what's wanted?

Reply:
Yes good point, the logic wasn't checking the "only" part before

Reply:
Any thoughts as to which of these is "right" ?

Reply:
I think the "only" can be removed. All the comment is saying is that we can't find a solution as long as this meta depends on a variable.

Reply:
So if we have a Meta with two constraints: Plus({Ext}, Meta1) and Plus({Ext2}, V) where Meta1 has a concrete solution and V is the graph variable...we should avoid solving it? Is that because solving it now mean we lose the ability to check the graph variable provides the same solution?

Reply:
If so, then yes, I'll remove the "only", and to be honest, the assert below seems pretty pointless now too (it exactly repeats the two if-conditions inside this loop over the same `self.extensions`).

I was thinking maybe the "only" was genuine, and if we can compute a solution without the variables, then we should - in which case branch `tests/infer_stack_overflow3`

Reply:
> So if we have a Meta with two constraints: Plus({Ext}, Meta1) and Plus({Ext2}, V) where Meta1 has a concrete solution and V is the graph variable...we should avoid solving it?

I think in that case we would've found the solution (Ext + get_solution(Meta1)) in `main_loop`, and would skip this step on that meta because a solution exists for it

Reply:
Ah, good point, yes thanks. Ok, will update comment + assert